### PR TITLE
[Core] QuickExit CoreWorker when GetCoreWorker is called after shutdown

### LINKED ISSUE #18995: [Bug] Worker exit crash
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

There is worker crash during exit:
```
[2021-09-30 05:19:46,712 I 16428 16428] core_worker.cc:145: Constructing CoreWorkerProcess. pid: 16428
[2021-09-30 05:19:46,715 I 16428 16428] grpc_server.cc:92: worker server started, listening on port 46733.
[2021-09-30 05:19:46,718 I 16428 16428] core_worker.cc:440: Initializing worker at address: 172.31.51.252:46733, worker ID 053e1985dbf8a0d8bfc5e1703fd59cd1061600ace1378d24ef3d4c67, raylet 6b1725ff964ab47c7fadd00c4527fea906ef714ac2547a4f7ddd3b07
[2021-09-30 05:19:46,828 I 16428 16428] io_service_pool.cc:35: IOServicePool is running with 1 io_service.
[2021-09-30 05:19:46,829 I 16428 16511] service_based_accessor.cc:611: Received notification for node id = 6b1725ff964ab47c7fadd00c4527fea906ef714ac2547a4f7ddd3b07, IsAlive = 1
[2021-09-30 05:19:50,186 I 16428 16511] actor_manager.cc:215: received notification on actor, state: ALIVE, actor_id: f6a23ffeae68356b95e5b6d901000000, ip address: 172.31.51.252, port: 43615, worker_id: 268170f2739fa80de6f4a39ff6f4fc409b5fcc84e8679cd287e0235e, raylet_id: 6b1725ff964ab47c7fadd00c4527fea906ef714ac2547a4f7ddd3b07, num_restarts: 0, has creation_task_exception=0
[2021-09-30 05:19:50,187 I 16428 16511] direct_actor_transport.cc:186: Connecting to actor f6a23ffeae68356b95e5b6d901000000 at worker 268170f2739fa80de6f4a39ff6f4fc409b5fcc84e8679cd287e0235e
[2021-09-30 05:19:50,189 I 16428 16428] core_worker.cc:2282: Finished executing task 158244e436517a1affffffffffffffffffffffff01000000, status=OK
[2021-09-30 05:19:51,229 I 16428 16511] core_worker.cc:775: Exit signal received, this process will exit after all outstanding tasks have finished, exit_type=IDLE_EXIT
[2021-09-30 05:19:51,229 I 16428 16428] core_worker.cc:327: Removed worker 053e1985dbf8a0d8bfc5e1703fd59cd1061600ace1378d24ef3d4c67
[2021-09-30 05:19:51,479 I 16428 16428] core_worker.cc:199: Destructing CoreWorkerProcess. pid: 16428
[2021-09-30 05:19:51,479 I 16428 16428] io_service_pool.cc:47: IOServicePool is stopped.
[2021-09-30 05:19:51,614 C 16428 16516] core_worker.cc:274:  Check failed: core_worker_process->global_worker_ global_worker_ must not be NULL
*** StackTrace Information ***
    ray::SpdLogMessage::Flush()
    ray::RayLog::~RayLog()
    ray::core::CoreWorkerProcess::GetCoreWorker()
    __pyx_pw_3ray_7_raylet_10CoreWorker_23get_worker_id()
    method_vectorcall_NOARGS
```

### Reproduction script

Run this script and it'll show up: https://github.com/ray-project/contrib-workflow-dag/blob/ml-codeflare/contrib/workflow/ml_codeflare/test_Datamodel.py

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?

There is a race condition between core worker shutdown versus language front end calling APIs on coreworkers. In this case we will crash as we failed the RAY_CHECK, which causes confusion to the users. 

We fix it by calling quick_exit if we know the core_worker is already shutdown.

## Related issue number

Closes #18995 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests: TBD, rerun the script in the
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]
I'm ok with this but it seems a bit strange. Is the proper solution to return an error code to the language layer?

We can add a TODO if this seems better.

**Comment by scv119** [comment]
yeah. a proper solution is return error code and handle it in language frontend. but it's a lot of work and ROI seems low the the moment.

**Comment by rkooo567** [comment]
What's the behavior for the end user when this happens? How would this propagate?

**Comment by rkooo567** [comment]
How would the error propagate to users? Let's say this happens in the driver, would this just abrubtly kill the driver?

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -89,6 +90,16 @@ ObjectLocation CreateObjectLocation(const rpc::GetObjectLocationsOwnerReply &rep
 /// The global instance of `CoreWorkerProcess`.
 std::unique_ptr<CoreWorkerProcess> core_worker_process;
 
+/// Teriminate the process without cleaning up the resources.
+/// It will flush the log if logging_enabled is set to true.
+void QuickExit(bool logging_enabled = false) {
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:95]: Isn't this enable_logging always true (when it is given)?

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -257,6 +258,18 @@ void CoreWorkerProcess::InitializeSystemConfig() {
   RayConfig::instance().initialize(promise.get_future().get());
 }
 
+bool CoreWorkerProcess::CreateGlobalWorkerOnConstruction() const {
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:261]: What about
```suggestion
bool CoreWorkerProcess::ShouldCreateGlobalWorkerOnConstruction() const {
```
?

File: src/ray/core_worker/core_worker.cc [hunk 3]
```diff
@@ -273,6 +286,14 @@ CoreWorker &CoreWorkerProcess::GetCoreWorker() {
   EnsureInitialized();
   if (core_worker_process->options_.num_workers == 1) {
     auto global_worker = core_worker_process->GetGlobalWorker();
+    if (core_worker_process->CreateGlobalWorkerOnConstruction() && !global_worker) {
+      // This could only happen when the worker has already been shutdown.
+      // In this case, we should exit without crashing.
+      // TODO (scv119): A better solution could be returning error code
+      // and handling it at language frontend.
+      RAY_LOG(ERROR) << "The global_worker has already been shut down.";
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:294]: Maybe improve the error message? "The global worker has already been shutdown. This happens when the Python code accesses the Ray's worker after it is shutdown. The process will exit"?

**Comment by scv119** [comment]
> How would the error propagate to users? Let's say this happens in the driver, would this just abruptly kill the driver?
I think in term of behavior there is no difference; before and after this change the process will exit when we enter into this condition. The difference is now we won't crash but just RAY_LOG(ERROR). Maybe the question is does this mask the real issue.

**Pull request merged**